run-tests.pl

```diff
@@ -527,9 +541,9 @@ sub _run_test
 
    my $success = eval {
       my @reqs;
-      my $f_setup = Future->needs_all( @req_futures )
+      my $f_setup = Future->needs_all( map { without_cancel($_) } @req_futures )
          ->on_done( sub { @reqs = @_ } )
-         ->on_fail( sub { die "fixture failed - $_[0]\n" } );
+         ->else( sub { die "fixture failed - $_[0]\n" } );
```

Comment:
This wants to stay as `on_fail` - it's not a piece of logic inserted *inline* into the failure-handling path, simply an incidental observer of the fact that failure happened.

Reply:
I'm not sure I follow. If it's an incidental observer, why do we want to `die` inside it? 

I thought the point of this code was that if any of the `@req_futures` fail, we want `$f_setup` to fail with a specific error message - so we *do* want to insert it into the failure-handling path.

Empirically, without this change, the test run just wedges (IIRC think it times out after 60s or something). The code in `on_fail` isn't `eval`ed, so the `die` ends up taking out lots of layers of the stack which seem to upset Future.pm. (It took me quite a while to track this down...)

Reply:
Having poked at this a bit harder, I think I see what you mean: I think you are hoping that the `die` will cause us to leave the `eval` block at https://github.com/matrix-org/sytest/blob/6c464959c614ac92d3055fa70b385a15e19d45c6/run-tests.pl#L542.

Unfortunately, the `die` is happening within the context of the `on_fail` callback rather than directly in that eval block, which means it takes out lots of intermediate layers of stack which might still be important.

By turning this into an `else`, we instead make the `get` at https://github.com/matrix-org/sytest/blob/6c464959c614ac92d3055fa70b385a15e19d45c6/run-tests.pl#L586 throw, which has the desired effect.